crates/matrix-sdk-crypto/src/machine.rs

```diff
@@ -1215,23 +1213,27 @@ impl OlmMachine {
 
     pub(crate) async fn preprocess_sync_changes(
         &self,
-        transaction: &mut StoreTransaction,
         sync_changes: EncryptionSyncChanges<'_>,
-    ) -> OlmResult<(Vec<Raw<AnyToDeviceEvent>>, Changes)> {
+    ) -> OlmResult<(StoreTransaction, Vec<Raw<AnyToDeviceEvent>>, Changes)> {
         // Remove verification objects that have expired or are done.
         let mut events = self.inner.verification_machine.garbage_collect();
 
-        // The account is automatically saved by the store transaction created by the
-        // caller.
-        let mut changes = Default::default();
+        // Updating key counts may generate new one-time keys, and we should *never*
+        // forget about them once generated (otherwise this leads to issues like
+        // #1415), so save them in a separate transaction here.
+        self.inner
+            .store
+            .with_transaction(|mut transaction| async {
+                let account = transaction.account().await.map_err(OlmError::Store)?;
+                account.update_key_counts(
+                    sync_changes.one_time_keys_counts,
+                    sync_changes.unused_fallback_keys,
+                );
+                Ok((transaction, ()))
+            })
+            .await?;
```

Comment:
I'm afraid I'm failing to follow how this fixes the problem.

Is there not still a possibility that committing the transaction can fail, leaving us in the same situation as before? (As I understand it, that situation is that the  `Account` has been updated, but not persisted to the database.)

Reply:
Hmm that's correct that this transaction can fail, if the account cannot be fetched. The rest of the code cannot fail.

Reply:
One solution would be to fetch the account before running the transaction maybe (if possible, because for now, the account is fetched by using the transaction).

Reply:
We haven't quite identified the cause of the problem yet, so we're still navigating a bit randomly here, and two days ago we *thought* this could be the source, but it's probably not.

But the *only* way to modify an `Account` is to get a transaction, that gives us a write-access scope to the account. There's no other way to upsert `Account`s in theory; in practice, that needs to be double-checked because I'm not sure the API to write `Account` to the database is private/hidden/limited.